dll/win32/shell32/wine/shellpath.c

```diff
@@ -474,42 +595,118 @@ int WINAPI PathCleanupSpec( LPCWSTR lpszPathW, LPWSTR lpszFileW )
 }
 
 /*************************************************************************
- * PathQualifyA		[SHELL32]
+ * PathQualifyW		[SHELL32]
  */
-static BOOL PathQualifyA(LPCSTR pszPath)
+static VOID PathQualifyW(LPWSTR pszPath)
 {
-	FIXME("%s\n",pszPath);
-	return FALSE;
+    TRACE("%s\n",debugstr_w(pszPath));
+    PathQualifyExW(pszPath, NULL, 0);
 }
 
 /*************************************************************************
- * PathQualifyW		[SHELL32]
+ * PathQualifyA		[SHELL32]
  */
-static BOOL PathQualifyW(LPCWSTR pszPath)
+static VOID PathQualifyA(LPSTR pszPath)
 {
-	FIXME("%s\n",debugstr_w(pszPath));
-	return FALSE;
+    WCHAR szPath[MAX_PATH];
+    TRACE("%s\n",pszPath);
+    SHAnsiToUnicode(pszPath, szPath, _countof(szPath));
+    PathQualifyW(szPath);
+    SHUnicodeToAnsi(szPath, pszPath, MAX_PATH);
 }
 
 /*************************************************************************
  * PathQualify	[SHELL32.49]
  */
-BOOL WINAPI PathQualifyAW(LPCVOID pszPath)
+VOID WINAPI PathQualifyAW(LPVOID pszPath)
 {
-	if (SHELL_OsIsUnicode())
-	  return PathQualifyW(pszPath);
-	return PathQualifyA(pszPath);
+    if (SHELL_OsIsUnicode())
+        PathQualifyW(pszPath);
+    else
+        PathQualifyA(pszPath);
 }
 
-static BOOL PathResolveA(LPSTR path, LPCSTR *paths, DWORD flags)
+BOOL WINAPI PathResolveA(LPSTR path, LPCSTR *dirs, DWORD flags)
 {
-    FIXME("(%s,%p,0x%08x),stub!\n", debugstr_a(path), paths, flags);
+    FIXME("(%s,%p,0x%08x),stub!\n", debugstr_a(path), dirs, flags);
     return FALSE;
 }
 
-static BOOL PathResolveW(LPWSTR path, LPCWSTR *paths, DWORD flags)
+BOOL WINAPI PathResolveW(LPWSTR path, LPCWSTR *dirs, DWORD flags)
 {
-    FIXME("(%s,%p,0x%08x),stub!\n", debugstr_w(path), paths, flags);
+    DWORD dwWhich;
+    TRACE("PathResolveW(%s,%p,0x%08x)\n", debugstr_w(path), dirs, flags);
+
+    dwWhich = ((flags & PRF_DONTFINDLNK) ? 0xF : 0x3F);
+
+    if (flags & PRF_VERIFYEXISTS)
+        SetLastError(ERROR_FILE_NOT_FOUND);
+
+    PathUnquoteSpacesW(path);
+
+    if (PathIsRootW(path))
+    {
+        if ((path[0] == L'\\' && path[1] == 0) ||
+            PathIsUNCServerW(path) || PathIsUNCServerShareW(path))
+        {
+            if (flags & PRF_FIRSTDIRDEF)
+                PathQualifyExW(path, dirs[0], 0);
+            else
+                PathQualifyExW(path, NULL, 0);
+        }
+
+        if (flags & PRF_VERIFYEXISTS)
+            return PathFileExistsAndAttributesW(path, NULL);
+        return TRUE;
+    }
+    else if (PathIsFileSpecW(path))
+    {
+        if ((flags & PRF_TRYPROGRAMEXTENSIONS) && PathSearchOnExtensionsW(path, dirs, TRUE, dwWhich))
```

Comment:
You certainly should be more precise on this flag. `PRF_TRYPROGRAMEXTENSIONS` is a combination of `PRF_EXECUTABLE` and `PRF_VERIFYEXISTS`, so this means that if one just passes the `PRF_EXECUTABLE` flag (legitimate use), what would happen there? This code wouldn't run. (Similarly, below, for other usages of the flag.)
I'm not sure this is what you want...

Reply:
Case 1 is `((flags & PRF_TRYPROGRAMEXTENSIONS) && PathSearchOnExtensionsW(path, dirs, TRUE, dwWhich))`.
Case 2 is `((flags & PRF_TRYPROGRAMEXTENSIONS) == PRF_TRYPROGRAMEXTENSIONS && PathSearchOnExtensionsW(path, dirs, TRUE, dwWhich))`.
Case 3 is `((flags & PRF_EXECUTABLE) && PathSearchOnExtensionsW(path, dirs, TRUE, dwWhich))`.
Case 4 is `((flags & PRF_VERIFYEXISTS) && PathSearchOnExtensionsW(path, dirs, TRUE, dwWhich))`.

Case 1:
![case1](https://user-images.githubusercontent.com/2107452/123181868-28071580-d4c9-11eb-9d17-f7952ba8ddc0.png)
324 failures.

Case 2:
![case2](https://user-images.githubusercontent.com/2107452/123181870-289fac00-d4c9-11eb-951c-b3fc4b07fbba.png)
356 failures.

Case 3:
![case3](https://user-images.githubusercontent.com/2107452/123181873-29d0d900-d4c9-11eb-9f17-333bec400a83.png)
340 failures.

Case 4:
![case4](https://user-images.githubusercontent.com/2107452/123181862-26d5e880-d4c9-11eb-9674-6d36204bd116.png)
340 failures.

Reply:
Thus, Case 1 is the best.